pkg/mimirpb/prealloc_rw2.go

```diff
@@ -30,27 +34,50 @@ func FromWriteRequestToRW2Request(rw1 *WriteRequest, commonSymbols []string, off
 	defer reuseSymbolsTable(symbols)
 	symbols.ConfigureCommonSymbols(offset, commonSymbols)
 
+	var metadataMap map[string]*MetricMetadata
+	if len(rw1.Metadata) > 0 {
+		metadataMap = make(map[string]*MetricMetadata, len(rw1.Metadata))
```

Comment:
We have an opportunity to pool this map.
The entire RW2 conversion flow is not deeply optimized. I would rather append the pooling to https://github.com/grafana/mimir/pull/12329 rather than this PR.

I also don't think pooling would have high impact after measuring/profiling the above changes, and pooling adds an element of memory unsafety, so I've left it out for now. Non-metadata requests don't allocate any extra anyway.